File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/infrastructure/ValorantStoreHttpAdapter.java
package fr.huiitre.tools.modules.riot.valorant.infrastructure;

import fr.huiitre.tools.modules.riot.valorant.application.catalog.ports.ValorantStorePort;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.*;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ValorantStoreHttpAdapter implements ValorantStorePort {

    private static final String ENTITLEMENTS_URL = "https://entitlements.auth.riotgames.com/api/token/v1";
    private static final String STOREFRONT_URL_TEMPLATE = "https://pd.%s.a.pvp.net/store/v3/storefront/%s";
    
    private static final String VP_CURRENCY_ID = "85ad13f7-3d1b-5128-9eb2-7cd8ee0b5741";
    private static final String SKIN_TYPE_ID = "e7c63390-eda7-46e0-bb7a-a6abdacd2433";
    private static final String CLIENT_PLATFORM = "ew0KCSJwbGF0Zm9ybVR5cGUiOiAiUEMiLA0KCSJwbGF0Zm9ybU9TIjogIldpbmRvd3MiLA0KCSJwbGF0Zm9ybU9TVmVyc2lvbiI6ICIxMC4wLjE5MDQyLjEuMjU2LjY0Yml0IiwNCgkicGxhdGZvcm1DaGlwc2V0IjogIlVua25vd24iDQp9";

    private final RestTemplate restTemplate;

    public ValorantStoreHttpAdapter(RestTemplate restTemplate) {
        this.restTemplate = restTemplate;
    }

    @Override
    public String fetchEntitlementsToken(String accessToken) {
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(accessToken);
        headers.setContentType(MediaType.APPLICATION_JSON);

        HttpEntity<String> request = new HttpEntity<>("{}", headers);

        try {
            ResponseEntity<Map<String, Object>> response = restTemplate.exchange(
                    ENTITLEMENTS_URL, HttpMethod.POST, request,
                    new ParameterizedTypeReference<>() {});

            Map<String, Object> body = response.getBody();
            if (body == null || !body.containsKey("entitlements_token")) {
                throw new IllegalArgumentException("RIOT_ENTITLEMENTS_EMPTY_RESPONSE");
            }

            return (String) body.get("entitlements_token");
        } catch (HttpClientErrorException e) {
            throw new IllegalArgumentException("RIOT_ACCESS_TOKEN_INVALID");
        }
    }

    @Override
    public RawStorefront fetchStorefront(String puuid, String region, String accessToken, String entitlementsToken, String clientVersion) {
        String url = String.format(STOREFRONT_URL_TEMPLATE, region, puuid);
        
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(accessToken);
        headers.set("X-Riot-Entitlements-JWT", entitlementsToken);
        headers.set("X-Riot-ClientPlatform", CLIENT_PLATFORM);
        headers.set("X-Riot-ClientVersion", clientVersion);
        headers.setContentType(MediaType.APPLICATION_JSON);

        HttpEntity<String> request = new HttpEntity<>("{}", headers);

        try {
            ResponseEntity<Map<String, Object>> response = restTemplate.exchange(
                    url, HttpMethod.POST, request,
                    new ParameterizedTypeReference<>() {});

            Map<String, Object> body = response.getBody();
            if (body == null) throw new IllegalArgumentException("RIOT_STOREFRONT_EMPTY_RESPONSE");

            return parseStorefront(body);
        } catch (HttpClientErrorException e) {
            throw new IllegalArgumentException("RIOT_STOREFRONT_FETCH_FAILED");
        }
    }

    private RawStorefront parseStorefront(Map<String, Object> data) {
        // 1. Single Item Offers
        List<RawOffer> offers = new ArrayList<>();
        Map<String, Object> skinsPanel = (Map<String, Object>) data.get("SkinsPanelLayout");
        if (skinsPanel != null) {
            List<Map<String, Object>> singleOffers = (List<Map<String, Object>>) skinsPanel.get("SingleItemStoreOffers");
            if (singleOffers != null) {
                for (Map<String, Object> o : singleOffers) {
                    List<Map<String, Object>> rewards = (List<Map<String, Object>>) o.get("Rewards");
                    Map<String, Integer> cost = (Map<String, Integer>) o.get("Cost");
                    if (rewards != null && !rewards.isEmpty()) {
                        String itemId = (String) rewards.get(0).get("ItemID");
                        int price = cost != null ? cost.getOrDefault(VP_CURRENCY_ID, 0) : 0;
                        offers.add(new RawOffer(itemId, price));
                    }
                }
            }
        }
        long remainingOffers = skinsPanel != null ? ((Number) skinsPanel.getOrDefault("SingleItemOffersRemainingDurationInSeconds", 0L)).longValue() : 0;

        // 2. Bundles
        List<RawBundle> bundles = new ArrayList<>();
        Map<String, Object> featured = (Map<String, Object>) data.get("FeaturedBundle");
        if (featured != null) {
            List<Map<String, Object>> rawBundles = (List<Map<String, Object>>) featured.get("Bundles");
            if (rawBundles == null && featured.containsKey("Bundle")) {
                rawBundles = List.of((Map<String, Object>) featured.get("Bundle"));
            }
            
            if (rawBundles != null) {
                for (Map<String, Object> b : rawBundles) {
                    long bundleRemaining = ((Number) b.getOrDefault("DurationRemainingInSeconds", 
                            featured.getOrDefault("BundleRemainingDurationInSeconds", 0L))).longValue();
                    
                    List<RawOffer> bundleItems = new ArrayList<>();
                    List<Map<String, Object>> items = (List<Map<String, Object>>) b.get("Items");
                    if (items != null) {
                        for (Map<String, Object> i : items) {
                            Map<String, Object> itemMeta = (Map<String, Object>) i.get("Item");
                            if (itemMeta != null && SKIN_TYPE_ID.equals(itemMeta.get("ItemTypeID"))) {
                                String itemId = (String) itemMeta.get("ItemID");
                                int price = ((Number) i.getOrDefault("DiscountedPrice", i.getOrDefault("BasePrice", 0))).intValue();
                                bundleItems.add(new RawOffer(itemId, price));
                            }
                        }
                    }

                    Map<String, Integer> totalBaseCost = (Map<String, Integer>) b.get("TotalBaseCost");
                    Map<String, Integer> totalDiscountedCost = (Map<String, Integer>) b.get("TotalDiscountedCost");
                    
                    int base = totalBaseCost != null ? totalBaseCost.getOrDefault(VP_CURRENCY_ID, 0) : 0;
                    int disc = totalDiscountedCost != null ? totalDiscountedCost.getOrDefault(VP_CURRENCY_ID, base) : base;

                    bundles.add(new RawBundle(
                            (String) b.get("DataAssetID"),
                            bundleItems,
                            base,
                            disc,
                            ((Number) b.getOrDefault("TotalDiscountPercent", 0)).intValue(),
                            bundleRemaining
                    ));
                }
            }
        }

        // 3. Night Market
        RawNightMarket nightMarket = null;
        Map<String, Object> bonusStore = (Map<String, Object>) data.get("BonusStore");
        if (bonusStore != null) {
            List<RawNightMarketOffer> nmOffers = new ArrayList<>();
            List<Map<String, Object>> rawNmOffers = (List<Map<String, Object>>) bonusStore.get("BonusStoreOffers");
            if (rawNmOffers != null) {
                for (Map<String, Object> o : rawNmOffers) {
                    Map<String, Object> offer = (Map<String, Object>) o.get("Offer");
                    List<Map<String, Object>> rewards = (List<Map<String, Object>>) offer.get("Rewards");
                    Map<String, Integer> discCosts = (Map<String, Integer>) o.get("DiscountCosts");
                    Map<String, Integer> origCosts = (Map<String, Integer>) offer.get("Cost");
                    
                    String itemId = rewards != null && !rewards.isEmpty() ? (String) rewards.get(0).get("ItemID") : "";
                    int discPrice = discCosts != null ? discCosts.getOrDefault(VP_CURRENCY_ID, 0) : 0;
                    int origPrice = origCosts != null ? origCosts.getOrDefault(VP_CURRENCY_ID, 0) : 0;

                    nmOffers.add(new RawNightMarketOffer(
                            (String) o.get("BonusOfferID"),
                            itemId,
                            origPrice,
                            discPrice,
                            ((Number) o.getOrDefault("DiscountPercent", 0)).intValue(),
                            (Boolean) o.getOrDefault("IsSeen", false)
                    ));
                }
            }
            nightMarket = new RawNightMarket(nmOffers, ((Number) bonusStore.getOrDefault("BonusStoreRemainingDurationInSeconds", 0L)).longValue());
        }

        return new RawStorefront(offers, remainingOffers, bundles, nightMarket);
    }
}

File: api/Modules/Riot/Valorant/Infrastructure/ValorantStoreHttpAdapter.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Ports;

namespace Tools.Api.Modules.Riot.Valorant.Infrastructure;

// Boutique et pseudo, lus sur les points d'entrée du client de jeu.
//
// Riot n'expose pas ces routes publiquement : elles attendent les en-têtes que le client envoie,
// dont une empreinte de plateforme et la version exacte du client. Une version périmée fait
// répondre 400, d'où la lecture de version.json à chaque appel.
public sealed class ValorantStoreHttpAdapter(HttpClient httpClient) : IValorantStorePort
{
    private const string EntitlementsUrl = "https://entitlements.auth.riotgames.com/api/token/v1";
    private const string StorefrontUrlTemplate = "https://pd.{0}.a.pvp.net/store/v3/storefront/{1}";
    private const string NameServiceUrlTemplate = "https://pd.{0}.a.pvp.net/name-service/v2/players";

    // Valorant Points. Les autres devises du panier (Radianite, Kingdom Credits) ne nous servent pas.
    private const string VpCurrencyId = "85ad13f7-3d1b-5128-9eb2-7cd8ee0b5741";

    // ItemTypeID « EquippableSkinLevel » : le seul type d'objet que le catalogue local sait résoudre.
    private const string SkinTypeId = "e7c63390-eda7-46e0-bb7a-a6abdacd2433";

    // Empreinte de plateforme attendue par Riot, encodée en Base64 (PC / Windows 10).
    private const string ClientPlatform =
        "ew0KCSJwbGF0Zm9ybVR5cGUiOiAiUEMiLA0KCSJwbGF0Zm9ybU9TIjogIldpbmRvd3MiLA0KCSJwbGF0Zm9ybU9TVmVyc2lvbiI6ICIxMC4wLjE5MDQyLjEuMjU2LjY0Yml0IiwNCgkicGxhdGZvcm1DaGlwc2V0IjogIlVua25vd24iDQp9";

    public async Task<string> FetchEntitlementsToken(string accessToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, EntitlementsUrl)
        {
            Content = EmptyJsonBody()
        };
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);

        var payload = await Send(request, "RIOT_ACCESS_TOKEN_INVALID");

        return payload.TryGetProperty("entitlements_token", out var token) && token.GetString() is { Length: > 0 } value
            ? value
            : throw AppException.Unavailable(
                "RIOT_ENTITLEMENTS_EMPTY_RESPONSE",
                "Riot n'a pas renvoyé de jeton d'entitlements.");
    }

    public async Task<IValorantStorePort.RawStorefront> FetchStorefront(
        string puuid,
        string region,
        string accessToken,
        string entitlementsToken,
        string clientVersion)
    {
        var url = string.Format(StorefrontUrlTemplate, region, puuid);

        using var request = new HttpRequestMessage(HttpMethod.Post, url) { Content = EmptyJsonBody() };
        AddGameClientHeaders(request, accessToken, entitlementsToken, clientVersion);

        return ParseStorefront(await Send(request, "RIOT_STOREFRONT_FETCH_FAILED"));
    }

    // Le pseudo n'est qu'un confort d'affichage : un échec rend un Riot ID vide plutôt qu'une erreur.
    public async Task<IValorantStorePort.RiotId> FetchRiotId(
        string puuid,
        string region,
        string accessToken,
        string entitlementsToken,
        string clientVersion)
    {
        var url = string.Format(NameServiceUrlTemplate, region);

        using var request = new HttpRequestMessage(HttpMethod.Put, url)
        {
            Content = JsonContent.Create(new[] { puuid })
        };
        AddGameClientHeaders(request, accessToken, entitlementsToken, clientVersion);

        try
        {
            var response = await httpClient.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                return new IValorantStorePort.RiotId(null, null);
            }

            var players = await response.Content.ReadFromJsonAsync<JsonElement>();

            if (players.ValueKind != JsonValueKind.Array || players.GetArrayLength() == 0)
            {
                return new IValorantStorePort.RiotId(null, null);
            }

            var player = players[0];

            return new IValorantStorePort.RiotId(
                ReadOptionalString(player, "GameName"),
                ReadOptionalString(player, "TagLine"));
        }
        catch (Exception exception) when (exception is HttpRequestException or TaskCanceledException or JsonException)
        {
            return new IValorantStorePort.RiotId(null, null);
        }
    }

    private static StringContent EmptyJsonBody() => new("{}", Encoding.UTF8, "application/json");

    private static void AddGameClientHeaders(
        HttpRequestMessage request,
        string accessToken,
        string entitlementsToken,
        string clientVersion)
    {
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
        request.Headers.Add("X-Riot-Entitlements-JWT", entitlementsToken);
        request.Headers.Add("X-Riot-ClientPlatform", ClientPlatform);
        request.Headers.Add("X-Riot-ClientVersion", clientVersion);
    }

    // clientErrorCode est celui que GetValorantStoreUseCase reconnaît pour retenter une fois avec
    // un access token neuf : un 4xx ici veut presque toujours dire « jeton expiré ».
    private async Task<JsonElement> Send(HttpRequestMessage request, string clientErrorCode)
    {
        HttpResponseMessage response;

        try
        {
            response = await httpClient.SendAsync(request);
        }
        catch (Exception exception) when (exception is HttpRequestException or TaskCanceledException)
        {
            throw AppException.Unavailable("RIOT_STORE_UNAVAILABLE", $"Riot est injoignable : {exception.Message}");
        }

        if ((int)response.StatusCode is >= 400 and < 500)
        {
            throw AppException.Validation(clientErrorCode, "Riot a refusé la requête de boutique.");
        }

        if (!response.IsSuccessStatusCode)
        {
            throw AppException.Unavailable(
                "RIOT_STORE_UNAVAILABLE",
                $"Riot a répondu {(int)response.StatusCode}.");
        }

        try
        {
            return await response.Content.ReadFromJsonAsync<JsonElement>();
        }
        catch (JsonException exception)
        {
            throw AppException.Unavailable("RIOT_STORE_UNAVAILABLE", $"Réponse Riot illisible : {exception.Message}");
        }
    }

    private static IValorantStorePort.RawStorefront ParseStorefront(JsonElement data)
    {
        var offers = new List<IValorantStorePort.RawOffer>();
        long remainingSeconds = 0;

        if (data.TryGetProperty("SkinsPanelLayout", out var skinsPanel))
        {
            remainingSeconds = ReadInt64(skinsPanel, "SingleItemOffersRemainingDurationInSeconds");

            foreach (var offer in Array(skinsPanel, "SingleItemStoreOffers"))
            {
                if (Array(offer, "Rewards").FirstOrDefault() is { ValueKind: JsonValueKind.Object } reward
                    && ReadOptionalString(reward, "ItemID") is { } itemId)
                {
                    offers.Add(new IValorantStorePort.RawOffer(itemId, ReadCurrency(offer, "Cost")));
                }
            }
        }

        return new IValorantStorePort.RawStorefront(
            offers,
            remainingSeconds,
            ParseBundles(data),
            ParseNightMarket(data));
    }

    private static List<IValorantStorePort.RawBundle> ParseBundles(JsonElement data)
    {
        var bundles = new List<IValorantStorePort.RawBundle>();

        if (!data.TryGetProperty("FeaturedBundle", out var featured))
        {
            return bundles;
        }

        // Riot expose « Bundles » au pluriel, et parfois un unique « Bundle » : les deux formes
        // existent selon la rotation.
        var rawBundles = featured.TryGetProperty("Bundles", out var many) && many.ValueKind == JsonValueKind.Array
            ? many.EnumerateArray().ToList()
            : featured.TryGetProperty("Bundle", out var single) && single.ValueKind == JsonValueKind.Object
                ? [single]
                : [];

        foreach (var bundle in rawBundles)
        {
            var items = new List<IValorantStorePort.RawOffer>();

            foreach (var item in Array(bundle, "Items"))
            {
                if (!item.TryGetProperty("Item", out var meta)
                    || ReadOptionalString(meta, "ItemTypeID") != SkinTypeId
                    || ReadOptionalString(meta, "ItemID") is not { } itemId)
                {
                    continue;
                }

                // Le prix remisé prime ; sans lui, le prix de base. Un skin offert vaut 0.
                var price = item.TryGetProperty("DiscountedPrice", out var discounted)
                    ? ToInt32(discounted)
                    : ReadInt32(item, "BasePrice");

                items.Add(new IValorantStorePort.RawOffer(itemId, price));
            }

            var baseCost = ReadCurrency(bundle, "TotalBaseCost");

            bundles.Add(new IValorantStorePort.RawBundle(
                ReadOptionalString(bundle, "DataAssetID") ?? string.Empty,
                items,
                baseCost,
                bundle.TryGetProperty("TotalDiscountedCost", out _) ? ReadCurrency(bundle, "TotalDiscountedCost") : baseCost,
                ReadInt32(bundle, "TotalDiscountPercent"),
                bundle.TryGetProperty("DurationRemainingInSeconds", out var bundleRemaining)
                    ? ToInt64(bundleRemaining)
                    : ReadInt64(featured, "BundleRemainingDurationInSeconds")));
        }

        return bundles;
    }

    private static IValorantStorePort.RawNightMarket? ParseNightMarket(JsonElement data)
    {
        if (!data.TryGetProperty("BonusStore", out var bonusStore))
        {
            return null;
        }

        var nightMarketOffers = new List<IValorantStorePort.RawNightMarketOffer>();

        foreach (var bonusOffer in Array(bonusStore, "BonusStoreOffers"))
        {
            if (!bonusOffer.TryGetProperty("Offer", out var offer))
            {
                continue;
            }

            var itemId = Array(offer, "Rewards").FirstOrDefault() is { ValueKind: JsonValueKind.Object } reward
                ? ReadOptionalString(reward, "ItemID") ?? string.Empty
                : string.Empty;

            nightMarketOffers.Add(new IValorantStorePort.RawNightMarketOffer(
                ReadOptionalString(bonusOffer, "BonusOfferID") ?? string.Empty,
                itemId,
                ReadCurrency(offer, "Cost"),
                ReadCurrency(bonusOffer, "DiscountCosts"),
                ReadInt32(bonusOffer, "DiscountPercent"),
                bonusOffer.TryGetProperty("IsSeen", out var seen) && seen.ValueKind == JsonValueKind.True));
        }

        return new IValorantStorePort.RawNightMarket(
            nightMarketOffers,
            ReadInt64(bonusStore, "BonusStoreRemainingDurationInSeconds"));
    }

    // Les prix sont des dictionnaires indexés par devise : seul le montant en VP nous intéresse.
    private static int ReadCurrency(JsonElement element, string propertyName) =>
        element.TryGetProperty(propertyName, out var costs)
        && costs.ValueKind == JsonValueKind.Object
        && costs.TryGetProperty(VpCurrencyId, out var amount)
            ? ToInt32(amount)
            : 0;

    private static IEnumerable<JsonElement> Array(JsonElement element, string propertyName) =>
        element.TryGetProperty(propertyName, out var array) && array.ValueKind == JsonValueKind.Array
            ? array.EnumerateArray()
            : [];

    private static string? ReadOptionalString(JsonElement element, string propertyName) =>
        element.TryGetProperty(propertyName, out var property) ? property.GetString() : null;

    private static int ReadInt32(JsonElement element, string propertyName) =>
        element.TryGetProperty(propertyName, out var property) ? ToInt32(property) : 0;

    private static long ReadInt64(JsonElement element, string propertyName) =>
        element.TryGetProperty(propertyName, out var property) ? ToInt64(property) : 0;

    // **Riot ne renvoie pas que des entiers.** TotalDiscountPercent porte une partie décimale, et
    // `GetInt32()` lève une FormatException dès que le JSON n'est pas un entier — là où le Java
    // passait par `Number.intValue()`, qui tronque sans rien dire. On tronque donc pareil, sur
    // tous les nombres : une valeur inattendue ne doit pas faire tomber toute la boutique.
    private static int ToInt32(JsonElement element) =>
        element.ValueKind == JsonValueKind.Number
            ? element.TryGetInt32(out var integer) ? integer : (int)element.GetDouble()
            : 0;

    private static long ToInt64(JsonElement element) =>
        element.ValueKind == JsonValueKind.Number
            ? element.TryGetInt64(out var integer) ? integer : (long)element.GetDouble()
            : 0;
}
